### PR TITLE
DCOS-40046 - Non Leader Masters

### PR DESCRIPTION
**Pull request by Fabs**
image

## Testing

Please test this on a single master and a multi master cluster. You can try soak if you just wanna read data. To kill master you need to be able to ssh into a master node and do `systemctl stop dcos-mesos-master`. Please ask around if you need help with that.

## Trade-offs

**Single master problem**
@lilysamimi I had two ideas to resolve the problem of not having more than one master in a cluster. I ended up doing Idea 1 because it is simpler.

Idea 1
https://cl.ly/402E30092Y1Z

It is the simples idea, but I dont like that we have 3 masters, andthen we figure the leader, drop to 2. On a single node cluster you will never see the Non-Leader

Idea 2
Multi Master: https://cl.ly/381d2o0E180H
Single Master: https://cl.ly/2H0z031x1j0V

**TICKETS**
this ticket also closes DCOS-39853, that was a way I could finish everything before going into my vacation. I am sorry it makes it harder to review.

## Dependencies

~Depends on #3127~

### CODE REVIEW

**Comment by Fabs** [comment]
@lilysamimi can you please check the tradeoffs with Julian and make a call (or even say Idea 3...). I left Idea 2 implemented because it is easier to throw stuff away and move to Idea 1.

**Comment by Fabs** [comment]
Also @juliangieseke, CI is broken like last time :/

**Comment by lilysamimi** [comment]
@Fabs thanks for putting these options together! I like idea 2 better than idea 1. Though I am wondering why we need to show the non-leader section in a single master? 

What's idea 3?

**Comment by juliangieseke** [comment]
@nLight rebased & pushed 👍

File: plugins/nodes/src/js/components/NodesGridView.js [hunk 1]
```diff
@@ -86,8 +86,7 @@ var NodesGridView = React.createClass({
 
     // Add 'Others' node to the list
     if (activeServiceIds.length > MAX_SERVICES_TO_SHOW) {
-      var classNameOther =
-        `dot service-color-${OTHER_SERVICES_COLOR}`;
+      var classNameOther = `dot service-color-${OTHER_SERVICES_COLOR}`;
```
Comment (nLight) [inline plugins/nodes/src/js/components/NodesGridView.js]: shall we do `const` as we're touching the line anyway?

Comment (juliangieseke) [reply plugins/nodes/src/js/components/NodesGridView.js]: I'd not touch it, it made it to master some how and now our tooling changes it back in every PR, changing it in one would result in a ton of conflicts… :(

File: plugins/nodes/src/js/components/NonLeadersGrid.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+const NonLeaderList = ({ masters }) => {
+  if (masters.length === 0) {
+    return <EmptyLeaderList />;
+  }
+
+  return (
+    <ConfigurationMapSection>
+      {masters.map(master => (
```
[the remark this answers is not in the document]
Comment (nLight) [reply plugins/nodes/src/js/components/NonLeadersGrid.js]: not me

File: plugins/nodes/src/js/data/MesosMasters.js [hunk 3]
```diff
@@ -1,84 +1,83 @@
...
+  if (masters === undefined) {
+    return undefined;
+  }
 
-export function getRegion(master) {
-  return (
-    findNestedPropertyInObject(master, "domain.fault_domain.region.name") ||
-    "N/A"
-  );
+  return masters.filter(master => master.host_ip !== leader.hostIp);
```
Comment (nLight) [inline plugins/nodes/src/js/data/MesosMasters.js]: There's no way of aligning `host_ip` and `hostIp`, is there?

File: plugins/nodes/src/js/data/MesosMastersHealth.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    Object.assign({}, master, {
+      healthDescription: UnitHealthUtil.getHealth(parseInt(master.health, 10))
+    })
+  );
+}
+
+export function mesosMastersHealthQuery(source, interval) {
+  const request$ = source().map(withHealthDescription);
+
+  return replayStream(pollStream(interval, request$));
```
Comment (nLight) [inline plugins/nodes/src/js/data/MesosMastersHealth.js]: How would you like having `request$.pollStream(interval).replayStream()`? It should be possible with `lift`, though a bit of overhead could be involved.

File: plugins/nodes/src/js/data/MesosMastersHealth.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+const HEALTH_POLL_INTERVAL = 30000;
```
Comment (nLight) [inline plugins/nodes/src/js/data/MesosMastersHealth.js]: Noticed that in this file the const is right beside the function while in MesosMastersLeader.js `const STORE_POLL_INTERVAL` is at the beginning of the file. How would you feel aligning those?

File: plugins/nodes/src/js/data/MesosMastersLeader.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import "rxjs/add/operator/retry";
+import "rxjs/add/operator/filter";
+
+import MesosStateStore from "#SRC/js/stores/MesosStateStore";
+import { findNestedPropertyInObject } from "#SRC/js/utils/Util";
+
+const STORE_POLL_INTERVAL = 2000;
+
+function isEmptyObject(obj) {
+  return Object.keys(obj).length === 0;
```
Comment (nLight) [inline plugins/nodes/src/js/data/MesosMastersLeader.js]: we have `ValidatorUtil.isEmpty` would it do the trick in your case?

File: plugins/nodes/src/js/components/NodesGridView.js (hunk 1 above)
Comment (DanielMSchmidt) [reply plugins/nodes/src/js/components/NodesGridView.js]: maybe we should improve our prettier tooling, it seems to forget some files.

File: plugins/nodes/src/js/data/MesosMastersLeader.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+function hostPort(address) {
```
Comment (nLight) [inline plugins/nodes/src/js/data/MesosMastersLeader.js]: isn't it rather `hostWithPort`?

File: plugins/nodes/src/js/data/MesosMastersLeader.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  );
+}
+
+export function mesosMasterLeaderQuery(
+  masterDataSource,
+  interval = STORE_POLL_INTERVAL
+) {
+  return Observable.timer(0, interval)
+    .map(_ => {
+      return masterDataSource();
```
Comment (nLight) [inline plugins/nodes/src/js/data/MesosMastersLeader.js]: In the next `map` the one-line version is used, shall we do the same here?

File: plugins/nodes/src/js/components/LeaderGrid.js [hunk 9]
```diff
@@ -31,7 +31,7 @@ const ConfigurationRow = ({ keyValue, title, value }) => {
   );
 };
 
-export default function LeaderGrid({ master }) {
+export default function LeaderGrid({ leader }) {
```
Comment (DanielMSchmidt) [inline plugins/nodes/src/js/components/LeaderGrid.js]: Should we rename this to MasterGrid then?

File: plugins/nodes/src/js/components/NodesGridView.js (hunk 1 above)
Comment (nLight) [reply plugins/nodes/src/js/components/NodesGridView.js]: I'd love to trace the leak and see how we can prevent both `var` and non-prettified changes on master.

File: plugins/nodes/src/js/data/MesosMasters.js [hunk 10]
```diff
@@ -1,84 +1,83 @@
...
-      };
-    });
-}
-
 // This is an attempt of the mediator pattern without componentFromStream
-export default class MesosMasters extends React.Component {
-  constructor() {
-    super(...arguments);
+export function connectMasterComponent(initialState, stream) {
+  class MesosMasters extends React.Component {
```
Comment (DanielMSchmidt) [inline plugins/nodes/src/js/data/MesosMasters.js]: Why not use componentFromStream?

File: plugins/nodes/src/js/data/__tests__/MesosMasters-test.js [hunk 11]
```diff
@@ -1,70 +1,105 @@
...
-      })
-    );
-  });
-
-  describe("#getRegion", function() {
-    it("retrieves region correctly", function() {
-      const master = masterData(faultDomainData()).master_info;
+const TestableMasterComponent = ({ component }) => {
+  return (
+    <IntlProvider locale="en" messages={enUS}>
```
Comment (DanielMSchmidt) [inline plugins/nodes/src/js/data/__tests__/MesosMasters-test.js]: Why use the intl provider here?

File: plugins/nodes/src/js/components/NodesGridView.js (hunk 1 above)
Comment (DanielMSchmidt) [reply plugins/nodes/src/js/components/NodesGridView.js]: I think eslint is missing something, that's why I would love to move it into jest as a runner to have a central place where these file paths are stored

Comment (DanielMSchmidt) [reply plugins/nodes/src/js/components/NodesGridView.js]: But that's a completely different topic

**Comment by juliangieseke** [comment]
@nLight @Fabs is on vacation this week and I'm maintaining his PR.
He most probably will not be able to disscuss your comments - since you didnt block this PR, is there anything you want to be changed? if so, please mark them as blocker and I'll change it. 
If not, I'd love to see an approval :)

**Comment by nLight** [comment]
@juliangieseke I'm hesitating approving/blocking the PR. To me it feels uncomfortable having 855 LOC for such a perceivably small change. But since I'm completely not in this topic I decided to withhold my opinion.

File: plugins/nodes/src/js/components/NodesGridView.js (hunk 1 above)
Comment (juliangieseke) [reply plugins/nodes/src/js/components/NodesGridView.js]: @nLight @DanielMSchmidt sure, can one of you pick this up? 🙇

Comment (nLight) [reply plugins/nodes/src/js/components/NodesGridView.js]: on it

Comment (nLight) [reply plugins/nodes/src/js/components/NodesGridView.js]: https://github.com/dcos/dcos-ui/pull/3160

**Comment by juliangieseke** [comment]
adding hold merge label and extracting this from the feature branch - I'd like to have Fabs answering your comments (and I dont see any reason just merging this, we can finish it next sprint I'd say - please correct me if I miss something)

**Comment by juliangieseke** [comment]
rebased and this branch is broken _again_ - @Fabs why is this happening to your branches all the time? 😭lets investigate once you're back, maybe its something you;re doing slightly different that everyone else? 🧐

**Comment by Fabs** [comment]
I do `git rebase -i upstream` @juliangieseke. Could it be because is someone`s else branch, and not fabs branch?

I was used to have this kind of problems with feature branches and other people branches. I usually used git to tell me which commits where duplicate when I rebase, and then I dropped them.

**Comment by Fabs** [comment]
@lilysamimi was on vacation haskell land. 

I have no idea 3. The problem I am trying to deal with is that we do not know how many masters a cluster have beforehand. Which might cause the Non-Leader list to just appear showing loading, and then vanish because there are no non-leaders. I would like to keep the components on the screen in all cases, because then there are no surprises. For that I needed the message for when there are no more masters.

File: plugins/nodes/src/js/data/MesosMastersHealth.js (hunk 4 above)
Comment (Fabs) [reply plugins/nodes/src/js/data/MesosMastersHealth.js]: Which library will provide the `lift` function or do you propose another way to achieve that interface? I am happy enough that the functions compose (although I do not use the pipe or compose explicitly). Also, is `lift` maybe a confusing concept for the team?

I would love to try it but ATM I don't see how.

File: plugins/nodes/src/js/data/MesosMastersLeader.js (hunk 6 above)
Comment (Fabs) [reply plugins/nodes/src/js/data/MesosMastersLeader.js]: I think that would be way too much logic for what I need. I would like to have an isEmptyObject. But I don't feel like using the isEmpty on Validator

File: plugins/nodes/src/js/data/MesosMasters.js (hunk 10 above)
Comment (Fabs) [reply plugins/nodes/src/js/data/MesosMasters.js]: I wanna try something different more inline with @weblancaster ideas of connect. I though I could expend 10 minutes trying that.

**Comment by Fabs** [comment]
Yes. @juliangieseke, @DanielMSchmidt and @nLight. Ready for review again.

**Comment by lilysamimi** [comment]
@Fabs I think the message you have is great, just add a period at the end.

File: plugins/nodes/src/js/data/MesosMastersHealth.js (hunk 4 above)
Comment (nLight) [reply plugins/nodes/src/js/data/MesosMastersHealth.js]: Lift should be a part of RxJs. Could be an overhead in this particular case, though we could employ `lift` for our custom Observables.

Comment (Fabs) [reply plugins/nodes/src/js/data/MesosMastersHealth.js]: I looked on rxjs. I guess they have it internally but I was not able to find that on the docs. I found a pr from 2015 talking about that.

Comment (nLight) [reply plugins/nodes/src/js/data/MesosMastersHealth.js]: http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-lift

File: plugins/nodes/src/js/components/LeaderGrid.js (hunk 9 above)
Comment (Fabs) [reply plugins/nodes/src/js/components/LeaderGrid.js]: I thought this one was addressed. I think the name LeaderGrid and NonLeaderGrid makes more sense as a pair as MasterGrid and NonLeaderGrid.

**Pull request merged**